fms/models/gpt_oss.py

```diff
@@ -0,0 +1,647 @@
+        else:
+            kvheads = self.config.kvheads
+            assert self.config.nheads % self.config.kvheads == 0
+
+        emb_kq = self.config.head_dim
+        emb_v = self.config.head_dim
+
+        self.attn = MultiHeadAttention(
+            emb_dim=self.config.emb_dim,
+            emb_kq=emb_kq,
+            emb_v=emb_v,
+            nheads=self.config.nheads,
+            kvheads=kvheads,
+            p_dropout=self.config.p_dropout,
+            use_bias=True,
+            position_encoder=rotary_emb,
+            fused=self.config.fused_weights,
+            linear_config=self.config.linear_config,
+            scale_factor=None,
+            has_sinks=True,
+        )
+
+        if self.config.p_dropout != 0:
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+        self.ff_sub_layer = MOEFeedForward(
+            self.config.num_experts,
+            self.config.top_k_experts,
+            self.config.emb_dim,
+            self.config.emb_dim,
+            use_bias=True,
+            topk_first=True,
+            swiglu_limit=self.config.swiglu_limit,
+        )
+
+        if self.config.p_dropout != 0:
+            self.dropout = nn.Dropout(self.config.p_dropout)
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        *,
+        position_ids: Optional[torch.Tensor] = None,
+        past_key_value_state: Optional[Tuple[torch.Tensor,]] = None,
+        use_cache: bool = False,
+        **attn_kwargs: Unpack[AttentionKwargs],
+    ):
+        # if the cache is not empty, we need to get the kv cache for self and cross attention
+        self_attn_past_key_value = past_key_value_state
+        attn_kwargs["attn_name"] = attn_kwargs.get("attn_name", "sdpa_with_sinks")
+
+        # first we do MHA and Add&Norm
+        residual = x
+        x = self.ln(x)
+        x = self.attn(
+            q=x,
+            position_ids=position_ids,
+            past_key_value_state=self_attn_past_key_value,
+            use_cache=use_cache,
+            **attn_kwargs,
+        )
+        cache = None
+        if use_cache:
+            x, cache = x
+        if self.config.p_dropout != 0:
+            x = self.dropout(x)
+        # residual connection
+        x = x + residual
+
+        # then we do FF and Add&Norm
+        residual = x
+
+        x = self.ff_ln(x)
+        x = self.ff_sub_layer(x)
+        if self.config.p_dropout != 0:
+            x = self.dropout(x)
+        # another residual
+        x = x + residual
+
+        if use_cache:
+            return (x, cache)
+        else:
+            return x
+
+
+class GptOssHeadless(nn.Module):
+    def __init__(
+        self,
+        config: GptOssConfig,
+        distributed_strategy: DistributedStrategy = NoOpStrategy,
+        **kwargs,
+    ):
+        super(GptOssHeadless, self).__init__()
+        if config is not None:
+            self.config = config
+        else:
+            self.config = GptOssConfig
```

Comment:
nit: can simplify to `self.config = config or GptOssConfig`. But also by the `__init__` signature, `config` should never be `None`, so it's not clear why this code is needed.  Finally the last line should probably be `self.config = GptOssConfig()` with the class instantiated, right? If we need that line.